### ISSUE TITLE
Having some trouble publishing

### ISSUE THREAD

**Issue by abrenoch** [voryx/Thruway#235]
I'm having a rather difficult time getting thruway to interface with my crossbar.io server.. I seem to be able to connect to the server just fine, but trying to publish anything to it appears to yield no results (without throwing any kind of error).

Here is my php:
```
use Psr\Log\NullLogger;
use Thruway\ClientSession;
use Thruway\Connection;
use Thruway\Logging\Logger;

require __DIR__ . "/vendor/autoload.php";

error_reporting(E_ALL);

//Uncomment to disable logging
//Logger::set(new NullLogger());

$connection = new Connection(
    [
        "realm" => "realm1",
        "url"   => "ws://127.0.0.1:8080/ws"
    ]
);

$connection->on("open", function (ClientSession $session) use ($connection) {
        try {
            $session->publish("com.some.event", [], []);
        } catch (Exception $e) {
            echo $e->getMessage(), PHP_EOL;
            $connection->close();
        }

        echo "event published?", PHP_EOL;

        $connection->close();
    }
);

$connection->on("close", function ($reason) {
    die("The connected has closed with reason: {$reason}");
});

$connection->on("error", function ($reason) {
    echo "The connected has closed with error: {$reason}\n";
});

$connection->open();
```

And my crossbar.io config:
```
{
    "version": 2,
    "controller": {},
    "workers": [
        {
            "type": "router",
            "realms": [
                {
                    "name": "realm1",
                    "roles": [
                        {
                            "name": "anonymous",
                            "permissions": [
                                {
                                    "uri": "",
                                    "match": "prefix",
                                    "allow": {
                                        "call": true,
                                        "register": true,
                                        "publish": true,
                                        "subscribe": true
                                    },
                                    "disclose": {
                                        "caller": true,
                                        "publisher": true
                                    },
                                    "cache": true
                                }
                            ]
                        }
                    ]
                }
            ],
            "transports": [
                {
                    "type": "web",
                    "endpoint": {
                        "type": "tcp",
                        "port": 8080
                    },
                    "paths": {
                        "/": {
                            "type": "static",
                            "directory": "../web"
                        },
                        "ws": {
                            "type": "websocket"
                        }
                    }
                }
            ]
        }
    ]
}
```

Which produces the following output:
```
2017-03-03T13:19:18.5652340 notice     Changing PHP precision from 14 to 16
2017-03-03T13:19:18.5662340 info       [Thruway\Peer\Client 10476] New client created
2017-03-03T13:19:18.5662340 info       [Thruway\Transport\PawlTransportProvider 10476] Starting Transport
2017-03-03T13:19:18.5742340 info       [Thruway\Transport\PawlTransportProvider 10476] Pawl has connected
2017-03-03T13:19:18.5772350 info       [Thruway\Peer\Client 10476] We have been welcomed...
event published?
2017-03-03T13:19:18.5782350 info       [Thruway\Transport\PawlTransportProvider 10476] Pawl has closed
The connected has closed with reason: close
```

I am using autobahn.js for my client-side code which works perfectly for me, but when I try to publish `com.some.event` from thruway I see simply nothing from either autobahn or thruway.

I have also noticed that if I rename `$session->publish("com.some.event")` to something like `$session2->publish("com.some.event")` (purposely calling $session2 while it is not defined), the script does not seem to recognize the error (even within a try/catch statement). This might be by design, but it seemed strange to me.

Can anybody give me some advice here? This seems like it would be relatively straightforward but I just can't seem to get it to work!

Thanks!

**Comment by andrewminerd** [comment]
Publish is an asynchronous operation (as is almost everything in the React world), so I think you're just closing the connection before you've given it a chance to actually publish your event. Try something like:

```
$session->publish("com.some.event", [])->then(function() use ($connection) {
    echo "event published", PHP_EOL;
    $connection->close();
});
```

**Comment by abrenoch** [comment]
Hey thanks for the reply!

I actually _just_ figured out what was happening here (although your suggestion was certainly at the top of my suspicions)!

What I was doing wrong was actually in autobahn.js and not thruway. Autobahn allows indicating 'prefixes' for different topics, but I seem to have misunderstood what is actually happening behind the scenes there. Those prefixes do not seem to apply to the `publish` & `subscribe` methods, so I simply was misunderstanding what topic was actually being called.

So I had (for example) `session.prefix('test', 'com.app.test')`, and had incorrectly assumed that when I both subscribed and published 'test' it was really using 'com.app.test'.. But it was simply using 'test' as the topic.

I would have bet a foot that I tried thruway using 'test' as well as 'com.app.test', but maybe there was something else wrong when I tried that.

So in conclusion; I'm a dummy :)

Thanks for the help though!
